sdk/core/core-https/src/nodeHttpsClient.ts

```diff
@@ -0,0 +1,330 @@
+      headers: request.headers.toJSON()
+    };
+    return options;
+  }
+}
+
+async function prepareFormData(formData: FormDataMap, request: PipelineRequest): Promise<void> {
+  const requestForm = new FormData();
+  for (const formKey of Object.keys(formData)) {
+    const formValue = formData[formKey];
+    if (Array.isArray(formValue)) {
+      for (const subValue of formValue) {
+        requestForm.append(formKey, subValue);
+      }
+    } else {
+      requestForm.append(formKey, formValue);
+    }
+  }
+
+  request.body = requestForm;
+  request.formData = undefined;
+  const contentType = request.headers.get("Content-Type");
+  if (contentType && contentType.indexOf("multipart/form-data") !== -1) {
+    request.headers.set(
+      "Content-Type",
+      `multipart/form-data; boundary=${requestForm.getBoundary()}`
+    );
+  }
+  try {
+    const contentLength = await new Promise<number>((resolve, reject) => {
+      requestForm.getLength((err, length) => {
+        if (err) {
+          reject(err);
+        } else {
+          resolve(length);
+        }
+      });
+    });
+    request.headers.set("Content-Length", contentLength);
+  } catch (e) {
+    // ignore setting the length if this fails
+  }
+}
+
+function getResponseHeaders(res: IncomingMessage): HttpHeaders {
+  const headers = createHttpHeaders();
+  for (const header of Object.keys(res.headers)) {
+    const value = res.headers[header];
+    if (Array.isArray(value)) {
+      if (value.length > 0) {
+        headers.set(header, value[0]);
+      }
+    } else if (value) {
+      headers.set(header, value);
+    }
+  }
+  return headers;
+}
+
+function getResponseStream(
+  stream: IncomingMessage,
+  headers: HttpHeaders,
+  skipDecompressResponse = false
+): NodeJS.ReadableStream {
+  if (skipDecompressResponse) {
+    return stream;
+  }
+
+  const contentEncoding = headers.get("Content-Encoding");
+  if (contentEncoding === "gzip") {
+    const unzip = zlib.createGunzip();
+    stream.pipe(unzip);
+    return unzip;
+  } else if (contentEncoding === "deflate") {
+    const inflate = zlib.createInflate();
+    stream.pipe(inflate);
+    return inflate;
+  }
+
+  return stream;
+}
+
+function streamToText(stream: NodeJS.ReadableStream): Promise<string> {
+  return new Promise<string>((resolve, reject) => {
+    const buffer: string[] = [];
+
+    stream.on("data", (chunk) => {
+      if (typeof chunk === "string") {
+        buffer.push(chunk);
+      } else {
+        buffer.push(chunk.toString());
+      }
+    });
+    stream.on("end", () => {
+      resolve(buffer.join(""));
+    });
+    stream.on("error", (e) => {
+      reject(
+        new RestError(`Error reading response as text: ${e.message}`, {
+          code: RestError.PARSE_ERROR
+        })
+      );
+    });
+  });
+}
+
+function getBodyLength(body: RequestBodyType): number | null {
+  if (!body) {
+    return 0;
+  } else if (typeof body === "string") {
+    return body.length;
```

Comment:
I'd thought the content-length is in bytes so this should be `body.byteLength` for strings?

Reply:
JS string length is always the byte length, isn't it? @bterlson as the master of encodings

Reply:
MDN says it's in UTF-16 code units. I don't know what that means.  However we've seen issues before in storage upload when string length is different from number of bytes.

Reply:
This is a very interesting issue, thanks for pinging me!

This should definitely return the number of bytes (octets) in the *encoded* body. Hopefully, this will usually be utf-8. Unfortunately, JS string length will give you UTF-16 code units (i.e. the number of 2-byte units comprising the string). These will only agree for non-BMP characters, and so I think this calculation is incorrect.

It seems a little strange to be doing anything other than accessing byteLength on the encoded buffer before it hits the wire.

Reply:
OK yeah, I looked at what node-fetch did to doublecheck this logic and we can't rely on string length. In those cases we should let node write it out instead of setting the length.